File: packages/kapi-format/src/annotation.ts
/**
 * @neokapi/kapi-format — annotation overlays.
 *
 * Annotations are non-authoritative analytical overlays on a
 * Block/Run graph. They describe properties of content (protected
 * terms, glossary matches, review status, MT confidence, QA flags,
 * translator notes) without changing how the content is stored,
 * edited, or served back. A tool that doesn't understand an
 * annotation type MUST ignore it and process the authoritative
 * content correctly.
 *
 * Annotations live as overlay files on disk under
 * `.kapi/collections/<name>/annotations/<producer-namespace>.overlays.jsonl`
 * (inside a kapi project) or as Session.PutOverlay calls keyed by
 * (kind, blockHash) when running through a BlockStore session.
 *
 * Each `.overlays.jsonl` file is JSON Lines. The first line is a header
 * record; the rest are annotation records. Multiple annotation
 * files can coexist without coordination — different producers own
 * different files.
 *
 * Lifecycle rules:
 *
 *   1. Annotations are non-authoritative. Consumers that don't
 *      understand a type skip its file entirely.
 *   2. Annotations are layered. Producers write independent files;
 *      no cross-file merge semantics.
 *   3. Annotations are derivable. Losing an annotation file costs
 *      only regeneration; the authoritative content in the
 *      `.kbf.json` blocks is unchanged.
 *   4. Annotations can become stale when blocks change. Validators
 *      detect orphans (anchors that no longer resolve) and flag
 *      them. Producers re-run to refresh.
 *   5. Annotations never mutate authoritative content. Producers
 *      are read-only on `documents/`.
 *
 * Inline annotations (markers that wrap a range inside the runs
 * sequence, the way XLIFF 2.0's <mrk> does) are a possible future
 * extension. The current spec is overlay-only.
 */

import type { Block, PluralForm, Run } from "./block.ts";

// ─── Annotation file shape ────────────────────────────────────────

/**
 * The top-level structure of an `annotations/*.overlays.jsonl` file after
 * all lines have been parsed. On disk, `header` is the first line
 * of the file and every line after it is one `Annotation`.
 */
export interface AnnotationFile {
  header: AnnotationFileHeader;
  annotations: Annotation[];
}

/**
 * The header record (first line of the .overlays.jsonl file). Identifies the
 * annotation type, versions it, and records which archive state it
 * was produced against.
 */
export interface AnnotationFileHeader {
  type: "header";
  /**
   * Namespaced annotation type, e.g. `@neokapi/term-detector`,
   * `acme/glossary-v2`, `bowrain/review-status`. No central registry
   * — consumers that understand a namespace consume the file;
   * consumers that don't skip it.
   */
  annotationType: string;
  /** Version of the annotation type's data shape (producer-owned). */
  annotationVersion: string;
  producer: {
    id: string;
    version: string;
  };
  /** ISO 8601 timestamp of when the file was generated. */
  created: string;
  /**
   * SHA-256 of the source archive's manifest.json at the time the
   * annotations were produced. Consumers compare this against the
   * current manifest hash to detect potentially-stale annotations.
   */
  targetArchive: string;
}

/**
 * One annotation record. Subsequent lines in the .overlays.jsonl file.
 */
export interface Annotation {
  type: "annotation";
  /** Stable within the file. Not required to be globally unique. */
  id: string;
  /** The block this annotation is about. */
  block: string;
  anchor: AnnotationAnchor;
  /**
   * Producer-specific payload. The framework imposes no schema on
   * this field; consumers that understand the annotation type know
   * how to read it.
   */
  data: unknown;
}

// ─── Anchor shapes ────────────────────────────────────────────────

/**
 * Where inside a block an annotation sits. Four shapes, discriminated by
 * `kind`.
 *
 * Positions are run-relative rather than offsets into flattened text, so a
 * boundary stays where it was put when a neighbouring run is rewritten and can
 * sit either side of a placeholder. They are pathed, so a position inside a
 * plural form or select case is addressable rather than approximated.
 */
export type AnnotationAnchor = BlockAnchor | RunAnchor | RangeAnchor | FormAnchor;

/**
 * The whole block. Used for block-level metadata like review status, MT
 * confidence, or "this block contains PII".
 */
export interface BlockAnchor {
  kind: "block";
}

/**
 * One `ph`, `pcOpen` or `sub` run. `path` names the run sequence and `runId`
 * the run within it, which is what lets a validator confirm the anchor still
 * points at the run it described.
 */
export interface RunAnchor {
  kind: "run";
  path: RunPath;
  runId: string;
}

/**
 * A half-open span of the run sequence at `path`. Used when a producer finds a
 * substring — a protected term, a URL, a named entity — that no typed run
 * already covers. A span may cross run boundaries, so a phrase running through
 * a bold segment is one range rather than three.
 */
export interface RangeAnchor {
  kind: "range";
  path: RunPath;
  /** Omitted when it is the zero position, which is what the wire form does. */
  start?: RunPos;
  end?: RunPos;
}

/**
 * One plural form or select case — "this 'few' form has been professionally
 * reviewed", "this 'female' case is flagged by QA".
 */
export interface FormAnchor {
  kind: "form";
  /** Path to the containing plural or select run. */
  path: RunPath;
  /**
   * Which form (for plural runs) or case (for select runs). For plural, must
   * be one of `PluralForm`. For select, any key of the run's `cases`.
   */
  key: string;
}

/**
 * A character boundary in a run sequence: an index into the sequence and a
 * rune offset into that run's text. A run carrying no text takes offset 0, and
 * an index equal to the sequence length is the boundary past the last run.
 */
export interface RunPos {
  run: number;
  offset?: number;
}

/**
 * A path through a block's nested runs structure. Empty path refers to
 * `Block.source` itself.
 */
export type RunPath = RunPathStep[];

/**
 * One step in a `RunPath`. Discriminated by shape:
 * - `number` — index into a `Run[]` sequence.
 * - `{ plural: PluralForm }` — step into a `plural` run's form.
 * - `{ select: string }` — step into a `select` run's case.
 */
export type RunPathStep = number | { plural: PluralForm } | { select: string };

// ─── Anchor resolution ────────────────────────────────────────────

/**
 * Result of resolving an anchor against a block. On success the resolved
 * entity; on failure a machine-readable reason.
 */
export type AnchorResolution =
  | { ok: true; kind: "block"; block: Block }
  | { ok: true; kind: "run"; run: Run }
  | { ok: true; kind: "range"; text: string; runs: Run[] }
  | { ok: true; kind: "form"; runs: Run[] }
  | {
      ok: false;
      reason:
        | "block-not-found"
        | "path-out-of-bounds"
        | "path-wrong-kind"
        | "run-id-mismatch"
        | "range-out-of-bounds"
        | "form-not-found";
    };

/**
 * Resolve an anchor against the block it belongs to. Mirrors ResolveAnchor in
 * core/kbf.
 */
export function resolveAnchor(block: Block, anchor: AnnotationAnchor): AnchorResolution {
  if (anchor.kind === "block") {
    return { ok: true, kind: "block", block };
  }

  const walked = walkPath(block.source, anchor.path);
  if (walked === null) return { ok: false, reason: "path-out-of-bounds" };
  const { runs: seq, run: landed } = walked;

  if (anchor.kind === "run") {
    const run = landed ?? seq.find((r) => runIdOf(r) === anchor.runId) ?? null;
    if (run === null) return { ok: false, reason: "run-id-mismatch" };
    const id = runIdOf(run);
    if (id === null) return { ok: false, reason: "path-wrong-kind" };
    if (id !== anchor.runId) return { ok: false, reason: "run-id-mismatch" };
    return { ok: true, kind: "run", run };
  }

  if (anchor.kind === "range") {
    if (!rangeInBounds(seq, anchor)) {
      return { ok: false, reason: "range-out-of-bounds" };
    }
    const runs = extractRuns(seq, anchor);
    return { ok: true, kind: "range", text: runsText(runs), runs };
  }

  // A form belongs to the run the path landed on: a plural run carries no id
  // of its own, so it is addressed by position.
  if (landed === null) return { ok: false, reason: "path-out-of-bounds" };
  if ("plural" in landed) {
    const form = landed.plural.forms[anchor.key as PluralForm];
    if (!form) return { ok: false, reason: "form-not-found" };
    return { ok: true, kind: "form", runs: form };
  }
  if ("select" in landed) {
    const caseRuns = landed.select.cases[anchor.key];
    if (!caseRuns) return { ok: false, reason: "form-not-found" };
    return { ok: true, kind: "form", runs: caseRuns };
  }
  return { ok: false, reason: "path-wrong-kind" };
}

interface WalkResult {
  /** The sequence the path addresses — `topRuns` for an empty path. */
  runs: Run[];
  /** The run an index step last landed on, which a form anchor is about. */
  run: Run | null;
}

/**
 * Walk a path through a run tree. Returns the sequence it addresses and the
 * run it landed on, or null if any step is out of bounds or the wrong kind.
 */
function walkPath(topRuns: Run[], path: RunPath): WalkResult | null {
  let currentRuns: Run[] = topRuns;
  let currentRun: Run | null = null;

  for (const step of path) {
    if (typeof step === "number") {
      if (step < 0 || step >= currentRuns.length) return null;
      currentRun = currentRuns[step];
    } else if ("plural" in step) {
      if (currentRun === null || !("plural" in currentRun)) return null;
      const form = currentRun.plural.forms[step.plural];
      if (!form) return null;
      currentRuns = form;
      currentRun = null;
    } else {
      if (currentRun === null || !("select" in currentRun)) return null;
      const caseRuns = currentRun.select.cases[step.select];
      if (!caseRuns) return null;
      currentRuns = caseRuns;
      currentRun = null;
    }
  }

  return { runs: currentRuns, run: currentRun };
}

/**
 * Rune length of a run's text, or 0 for a run that carries none.
 *
 * Code points, not UTF-16 units: an offset must count the same thing here as
 * it does in the model, or the two sides disagree about where a position is
 * the moment content leaves the Basic Multilingual Plane.
 */
function runTextLength(run: Run): number {
  return "text" in run ? Array.from(run.text).length : 0;
}

/**
 * A boundary as the wire gives it. Go writes `start` and `end` with
 * `omitzero`, so a range beginning at the first run carries no `start` key at
 * all: an absent position is the zero position, not a malformed anchor. The
 * resolver reads annotation files off disk, where no type is policing them, so
 * it takes them as they come.
 */
function posOf(pos: RunPos | undefined): RunPos {
  return pos ?? { run: 0 };
}

function offsetInBounds(runs: Run[], pos: RunPos): boolean {
  const offset = pos.offset ?? 0;
  if (offset < 0) return false;
  if (pos.run === runs.length) return offset === 0;
  if (pos.run < 0 || pos.run > runs.length) return false;
  const run = runs[pos.run];
  return "text" in run ? offset <= runTextLength(run) : offset === 0;
}

function rangeInBounds(runs: Run[], anchor: RangeAnchor): boolean {
  const start = posOf(anchor.start);
  const end = posOf(anchor.end);
  if (start.run < 0 || end.run < start.run) return false;
  if (end.run > runs.length) return false;
  return offsetInBounds(runs, start) && offsetInBounds(runs, end);
}

/**
 * The sub-sequence a range covers. Boundary text runs are split at their
 * offsets; a run carrying no text is atomic and comes through whole unless it
 * sits on the exclusive end.
 */
function extractRuns(runs: Run[], anchor: RangeAnchor): Run[] {
  const out: Run[] = [];
  const start = posOf(anchor.start);
  const end = posOf(anchor.end);
  const startOffset = start.offset ?? 0;
  const endOffset = end.offset ?? 0;

  for (let i = start.run; i <= end.run && i < runs.length; i++) {
    const run = runs[i];
    if ("text" in run) {
      const chars = Array.from(run.text);
      const from = i === start.run ? Math.min(startOffset, chars.length) : 0;
      const to = i === end.run ? Math.min(endOffset, chars.length) : chars.length;
      if (from < to) out.push({ text: chars.slice(from, to).join("") });
      continue;
    }
    if (i === end.run && endOffset === 0) continue;
    out.push(run);
  }
  return out;
}

function runsText(runs: Run[]): string {
  return runs.map((r) => ("text" in r ? r.text : "")).join("");
}

function runIdOf(run: Run): string | null {
  if ("ph" in run) return run.ph.id;
  if ("pcOpen" in run) return run.pcOpen.id;
  if ("sub" in run) return run.sub.id;
  return null;
}

// ─── Validation utilities ────────────────────────────────────────

/**
 * Check an annotation's anchor against a block and return an
 * error if it doesn't resolve. Intended for orphan detection when
 * a producer's annotations are loaded after the block they
 * reference has changed.
 */
export function validateAnchor(
  block: Block,
  annotation: Annotation,
): AnnotationValidationError | null {
  // The record names the block it is about, so validating it against one that
  // is not that block is a mismatch rather than a resolution failure.
  if (annotation.block !== block.id) {
    return {
      annotationId: annotation.id,
      blockId: annotation.block,
      reason: "block-not-found",
      message: messageFor("block-not-found", annotation),
    };
  }
  const result = resolveAnchor(block, annotation.anchor);
  if (result.ok) return null;
  return {
    annotationId: annotation.id,
    blockId: block.id,
    reason: result.reason,
    message: messageFor(result.reason, annotation),
  };
}

function messageFor(
  reason: (AnchorResolution & { ok: false })["reason"],
  annotation: Annotation,
): string {
  switch (reason) {
    case "block-not-found":
      return `annotation "${annotation.id}" targets block "${annotation.block}" which does not match`;
    case "path-out-of-bounds":
      return `annotation "${annotation.id}" path is out of bounds in block "${annotation.block}"`;
    case "path-wrong-kind":
      return `annotation "${annotation.id}" path lands on a run of the wrong kind for its anchor`;
    case "run-id-mismatch":
      return `annotation "${annotation.id}" resolves to a run whose id does not match the recorded id (possible orphan)`;
    case "range-out-of-bounds":
      return `annotation "${annotation.id}" range is out of bounds in the run sequence it addresses`;
    case "form-not-found":
      return `annotation "${annotation.id}" targets a plural form or select case that does not exist on the block`;
  }
}

export interface AnnotationValidationError {
  annotationId: string;
  blockId: string;
  reason: (AnchorResolution & { ok: false })["reason"];
  message: string;
}

File: core/kbf/annotation.go
package kbf

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	"github.com/neokapi/neokapi/core/model"
	"io"
	"strings"
)

// AnnotationFile is the in-memory form of an `annotations/*.overlays.jsonl`
// file: one header record plus zero or more annotation records.
type AnnotationFile struct {
	Header      AnnotationFileHeader
	Annotations []Annotation
}

// AnnotationFileHeader is the first line of a .overlays.jsonl file: identifies
// the annotation type, its producer, and the archive state the
// annotations were produced against.
type AnnotationFileHeader struct {
	Type              string             `json:"type"`
	AnnotationType    string             `json:"annotationType"`
	AnnotationVersion string             `json:"annotationVersion"`
	Producer          AnnotationProducer `json:"producer"`
	Created           string             `json:"created"`
	TargetArchive     string             `json:"targetArchive"`
}

// AnnotationProducer identifies the tool that wrote an annotation
// file.
type AnnotationProducer struct {
	ID      string `json:"id"`
	Version string `json:"version"`
}

// Annotation is one record in a .overlays.jsonl file (every line after the
// header).
type Annotation struct {
	Type string `json:"type"`
	ID   string `json:"id"`
	// Block is the block the annotation is about; Anchor says where inside it.
	Block  string `json:"block"`
	Anchor Anchor `json:"anchor"`
	// Data is the producer-specific payload. The framework imposes
	// no schema here; consumers that understand the annotation type
	// negotiate via AnnotationType.
	Data json.RawMessage `json:"data,omitempty"`
}

// ───────── anchor resolution ─────────

// AnchorResolveReason is a machine-readable reason an anchor didn't
// resolve. Mirrors the six reasons in
// packages/kapi-format/src/annotation.ts.
type AnchorResolveReason string

const (
	ReasonBlockNotFound    AnchorResolveReason = "block-not-found"
	ReasonPathOutOfBounds  AnchorResolveReason = "path-out-of-bounds"
	ReasonPathWrongKind    AnchorResolveReason = "path-wrong-kind"
	ReasonRunIDMismatch    AnchorResolveReason = "run-id-mismatch"
	ReasonRangeOutOfBounds AnchorResolveReason = "range-out-of-bounds"
	ReasonFormNotFound     AnchorResolveReason = "form-not-found"
)

// AnchorResolution is the result of ResolveAnchor. On success,
// exactly one of the *Target fields is populated according to
// Kind; on failure, Err is the machine-readable failure reason.
type AnchorResolution struct {
	OK   bool
	Kind AnchorKind
	Err  AnchorResolveReason

	// Populated on AnchorBlock success.
	BlockTarget *Block
	// Populated on AnchorRun success.
	RunTarget *Run
	// Populated on AnchorRange success.
	RangeRuns []Run
	RangeText string
	// Populated on AnchorForm success.
	FormRuns []Run
}

// ResolveAnchor resolves an annotation's anchor against the block it names,
// returning either the resolved entity or a machine-readable reason it did not.
// Mirrors resolveAnchor in packages/kapi-format/src/annotation.ts.
func ResolveAnchor(block *Block, anchor Anchor) AnchorResolution {
	if block == nil {
		return AnchorResolution{OK: false, Err: ReasonBlockNotFound}
	}

	if anchor.Kind == AnchorBlock {
		return AnchorResolution{OK: true, Kind: AnchorBlock, BlockTarget: block}
	}

	seq, landed, ok := walkPath(block.Source, anchor.Path)
	if !ok {
		return AnchorResolution{OK: false, Err: ReasonPathOutOfBounds}
	}

	switch anchor.Kind {
	case AnchorRun:
		run := landed
		if run == nil {
			for i := range seq {
				if seq[i].RunID() == anchor.RunID && anchor.RunID != "" {
					run = &seq[i]
					break
				}
			}
		}
		if run == nil {
			// The path resolved; nothing in the sequence carries that id.
			return AnchorResolution{OK: false, Err: ReasonRunIDMismatch}
		}
		id := run.RunID()
		if id == "" {
			return AnchorResolution{OK: false, Err: ReasonPathWrongKind}
		}
		if id != anchor.RunID {
			return AnchorResolution{OK: false, Err: ReasonRunIDMismatch}
		}
		return AnchorResolution{OK: true, Kind: AnchorRun, RunTarget: run}

	case AnchorRange:
		if !anchor.InBounds(seq) {
			return AnchorResolution{OK: false, Err: ReasonRangeOutOfBounds}
		}
		runs := anchor.ExtractRuns(seq)
		return AnchorResolution{
			OK: true, Kind: AnchorRange,
			RangeRuns: runs,
			RangeText: model.RunsText(runs),
		}

	case AnchorForm:
		if landed == nil {
			return AnchorResolution{OK: false, Err: ReasonPathOutOfBounds}
		}
		if landed.Plural != nil {
			form, has := landed.Plural.Forms[PluralForm(anchor.Key)]
			if !has {
				return AnchorResolution{OK: false, Err: ReasonFormNotFound}
			}
			return AnchorResolution{OK: true, Kind: AnchorForm, FormRuns: form}
		}
		if landed.Select != nil {
			caseRuns, has := landed.Select.Cases[anchor.Key]
			if !has {
				return AnchorResolution{OK: false, Err: ReasonFormNotFound}
			}
			return AnchorResolution{OK: true, Kind: AnchorForm, FormRuns: caseRuns}
		}
		return AnchorResolution{OK: false, Err: ReasonPathWrongKind}
	}

	return AnchorResolution{OK: false, Err: ReasonPathWrongKind}
}

// walkPath walks `path` through `top` and returns the run sequence it
// addresses — `top` itself for an empty path, otherwise the branch a plural or
// select step descended into — together with the run an index step last landed
// on, which is what a form anchor is about. The last return is false on any
// out-of-bounds or wrong-kind step.
func walkPath(top []Run, path RunPath) ([]Run, *Run, bool) {
	current := top
	var currentRun *Run
	for _, step := range path {
		switch step.Kind {
		case StepIndex:
			if step.Index < 0 || step.Index >= len(current) {
				return nil, nil, false
			}
			r := current[step.Index]
			currentRun = &r
		case StepPlural:
			if currentRun == nil || currentRun.Plural == nil {
				return nil, nil, false
			}
			form, has := currentRun.Plural.Forms[step.PluralForm]
			if !has {
				return nil, nil, false
			}
			current = form
			currentRun = nil
		case StepSelect:
			if currentRun == nil || currentRun.Select == nil {
				return nil, nil, false
			}
			caseRuns, has := currentRun.Select.Cases[step.SelectValue]
			if !has {
				return nil, nil, false
			}
			current = caseRuns
			currentRun = nil
		}
	}
	return current, currentRun, true
}

// AnnotationValidationError mirrors
// packages/kapi-format/src/annotation.ts's AnnotationValidationError.
type AnnotationValidationError struct {
	AnnotationID string
	BlockID      string
	Reason       AnchorResolveReason
	Message      string
}

// ValidateAnchor checks an annotation's anchor against a Block and
// returns a validation error if it doesn't resolve. Suitable for
// orphan-detection validators that process annotation files after
// blocks may have been re-extracted.
func ValidateAnchor(block *Block, ann Annotation) *AnnotationValidationError {
	// The record names the block it is about, so validating it against one that
	// is not that block is a mismatch rather than a resolution failure.
	if block == nil || ann.Block != block.ID {
		return &AnnotationValidationError{
			AnnotationID: ann.ID,
			BlockID:      ann.Block,
			Reason:       ReasonBlockNotFound,
			Message:      messageFor(ReasonBlockNotFound, ann),
		}
	}
	res := ResolveAnchor(block, ann.Anchor)
	if res.OK {
		return nil
	}
	return &AnnotationValidationError{
		AnnotationID: ann.ID,
		BlockID:      ann.Block,
		Reason:       res.Err,
		Message:      messageFor(res.Err, ann),
	}
}

func messageFor(reason AnchorResolveReason, ann Annotation) string {
	switch reason {
	case ReasonBlockNotFound:
		return fmt.Sprintf("annotation %q targets block %q which does not match", ann.ID, ann.Block)
	case ReasonPathOutOfBounds:
		return fmt.Sprintf("annotation %q path is out of bounds in block %q", ann.ID, ann.Block)
	case ReasonPathWrongKind:
		return fmt.Sprintf("annotation %q path lands on a run of the wrong kind for its anchor", ann.ID)
	case ReasonRunIDMismatch:
		return fmt.Sprintf("annotation %q resolves to a run whose id does not match the recorded id (possible orphan)", ann.ID)
	case ReasonRangeOutOfBounds:
		return fmt.Sprintf("annotation %q character range exceeds the target text run", ann.ID)
	case ReasonFormNotFound:
		return fmt.Sprintf("annotation %q targets a plural form or select case that does not exist on the block", ann.ID)
	}
	return fmt.Sprintf("annotation %q: %s", ann.ID, reason)
}

// ───────── annotation file I/O (.overlays.jsonl) ─────────

// DecodeAnnotationFile parses a JSON-Lines annotation overlay from r.
// The first non-empty line must be a header record; subsequent
// non-empty lines are annotation records.
func DecodeAnnotationFile(r io.Reader) (*AnnotationFile, error) {
	br := bufio.NewReader(r)
	var out AnnotationFile

	// Read header.
	header, err := readJSONLine(br)
	if err != nil {
		return nil, fmt.Errorf("kbf: read annotation header: %w", err)
	}
	if header == nil {
		return nil, errors.New("kbf: empty annotation file")
	}
	if err := json.Unmarshal(header, &out.Header); err != nil {
		return nil, fmt.Errorf("kbf: decode annotation header: %w", err)
	}
	if out.Header.Type != "header" {
		return nil, fmt.Errorf("kbf: annotation header has unexpected type %q", out.Header.Type)
	}

	// Read records.
	for {
		line, err := readJSONLine(br)
		if err != nil {
			return nil, fmt.Errorf("kbf: read annotation record: %w", err)
		}
		if line == nil {
			break
		}
		var ann Annotation
		if err := json.Unmarshal(line, &ann); err != nil {
			return nil, fmt.Errorf("kbf: decode annotation record: %w", err)
		}
		if ann.Type != "annotation" {
			return nil, fmt.Errorf("kbf: annotation record has unexpected type %q", ann.Type)
		}
		out.Annotations = append(out.Annotations, ann)
	}
	return &out, nil
}

// readJSONLine reads one line from br, skipping empty lines, and
// returns nil at EOF.
func readJSONLine(br *bufio.Reader) ([]byte, error) {
	for {
		line, err := br.ReadBytes('\n')
		if len(line) == 0 && err == io.EOF {
			return nil, nil
		}
		if err != nil && err != io.EOF {
			return nil, err
		}
		trimmed := strings.TrimRight(string(line), "\r\n")
		if trimmed == "" {
			if err == io.EOF {
				return nil, nil
			}
			continue
		}
		return []byte(trimmed), nil
	}
}

// EncodeAnnotationFile writes a JSON-Lines annotation overlay to w.
// Each line is compact JSON (no indentation) terminated by LF; this
// keeps the file grep-friendly and diff-friendly as required by RFC
// 0001.
func EncodeAnnotationFile(w io.Writer, f *AnnotationFile) error {
	if f == nil {
		return errors.New("kbf: encode nil annotation file")
	}
	// Header.
	headerLine := AnnotationFileHeader{
		Type:              "header",
		AnnotationType:    f.Header.AnnotationType,
		AnnotationVersion: f.Header.AnnotationVersion,
		Producer:          f.Header.Producer,
		Created:           f.Header.Created,
		TargetArchive:     f.Header.TargetArchive,
	}
	if err := writeJSONLine(w, headerLine); err != nil {
		return fmt.Errorf("kbf: write annotation header: %w", err)
	}
	for i := range f.Annotations {
		ann := f.Annotations[i]
		if ann.Type == "" {
			ann.Type = "annotation"
		}
		if err := writeJSONLine(w, ann); err != nil {
			return fmt.Errorf("kbf: write annotation record: %w", err)
		}
	}
	return nil
}

func writeJSONLine(w io.Writer, v any) error {
	buf, err := json.Marshal(v)
	if err != nil {
		return err
	}
	if _, err := w.Write(buf); err != nil {
		return err
	}
	_, err = w.Write([]byte{'\n'})
	return err
}
